scipy/linalg/basic.py

```diff
@@ -1125,3 +1125,110 @@ def pinvh(a, cond=None, rcond=None, lower=True, return_rank=False,
         return B, len(psigma_diag)
     else:
         return B
+
+def balance(A, permute=True, scale=True, separate = False, overwrite_a=False):
+    """
+    A wrapper around LAPACK's xGEBAL routine family for matrix balancing. 
+    
+    The balancing tries to equalize the row and column 1-norms by applying 
+    a similarity transformation such that the magnitude variation of the 
+    matrix entries is reflected to the scaling matrices. This is particularly 
+    useful for eigenvalue and matrix decompositions and often automatically 
+    performed. Hence, this function provides a convenience for manually 
+    balancing. 
+        
+    Moreover, if turned on, the matrix is first permuted to isolate the upper 
+    triangular parts of the matrix and only the remaining part is scaled. 
+    
+    The balanced matrix satisfies the following equality 
+    
+    .. math::
+         
+                        B = T^{-1} A T 
+
+    The algorithm is based on the well-known technique of [1]_ and has 
+    been modified to account for special cases. See [2]_ for details
+    which have been implemented since LAPACK v3.5.0. 
+    
+    Parameters
+    ----------
+    A : (n, n) array_like
+        Square data matrix for the balancing.
+    permute : bool
+        The selector to define whether permutation of A is also performed
+        prior to scaling.
+    scale : bool
+        The selector to turn on and off the scaling. If False, the matrix 
+        will not be scaled. 
+    separate : bool
+        This switches from returning a full matrix of the transformation
+        to a tuple of two separate 1D permutation and scaling arrays.
+    overwrite_a : bool
+        This is passed to xGEBAL directly. Essentially, overwrites the result
+        to the data. It might increase the space efficiency. See LAPACK manual 
+        for details. This is False by default. 
+
+    Returns
+    -------
+    B : (n, n) array
+        Balanced matrix 
+    T : (n, n) array
+        A possibly permuted diagonal matrix whose nonzero entries are 
+        integer powers of 2 to avoid numerical truncation errors. 
+    s, p : (n,) array
+        If ``separate`` keyword is set to True then instead of the matrix 
+        ``T`` above, the scaling and the permutation vector is given. 
+
+    .. versionadded:: 0.19.0
+    
+    References
+    ----------
+    .. [1] : B.N. Parlett and C. Reinsch, "Balancing a Matrix for 
+       Calculation of Eigenvalues and Eigenvectors", Numerische Mathematik, 
+       Vol.13(4), 1969, DOI:10.1007/BF02165404
+       
+    .. [2] : R. James, J. Langou, B.R. Lowery, "On matrix balancing and 
+       eigenvector computation", 2014, Available online: 
+       http://arxiv.org/abs/1401.5766
+       
```

Comment:
Adding an ``Examples`` section would be nice. Can be simple, something like:

    >>> from scipy import linalg
    >>> x = np.array([[1,2,0], [9,1,0.01], [1,2,10*np.pi]])
    >>> y, permscale = linalg.balance(x)
    >>> np.abs(x).sum(axis=0) / np.abs(x).sum(axis=1)
    array([ 3.66666667,  0.4995005 ,  0.91312162])
    >>> np.abs(y).sum(axis=0) / np.abs(y).sum(axis=1)   # 1-norms close to equal now
    array([ 1.10625   ,  0.90547703,  1.00011878])
    >>> permscale  # only powers of 2 (0.5 == 2^(-1))
    array([[  0.5,   0. ,   0. ],
           [  0. ,   1. ,   0. ],
           [  0. ,   0. ,  16. ]])